feat(transformer): support tagged template expression with `</script` transformation

Dunqing opened the pull request:
Implement a transformation plugin for tagged template expressions containing </script> tags to prevent script tag issues in browser environments.

  This plugin transforms tagged template literals containing `</script` (case-insensitive) to use a helper function call with template caching, matching esbuild's behavior.

 ### Example

  Transforms tagged templates to use cached template objects with a helper function:

  Input:
```js
foo`</script>`
bar`<script>${content}</script>`
```
  Output:
```js
var _templateObject, _templateObject2;
foo(_templateObject || (_templateObject = babelHelpers.taggedTemplateLiteral(["<\/script>"])));
bar(_templateObject2 || (_templateObject2 = babelHelpers.taggedTemplateLiteral(["<script>", "<\/script>"])), content);
```

 Closes #15306

Dunqing commented:
* **#15664** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15664?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15664?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15664).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on npm/runtime/src/helpers/esm/taggedTemplateLiteralEscape.js:
```diff
@@ -0,0 +1,6 @@
+function _taggedTemplateLiteralEscape(e, t) {
+  return t || (t = e.slice(0)), Object.freeze(Object.defineProperty(e, "raw", {
+    value: Object.freeze(t)
+  }));
```
esbuild's helper function is like this
```suggestion
  return Object.freeze(Object.defineProperty(e, "raw", {
    value: Object.freeze(t || e.slice())
  }));
```
I think we can use it instead of the babel's one as this would be slightly shorter when minified.

sapphi-red replied:
Ah, ok. It feels a bit verbose, but it's definitely not a blocker.

Dunqing commented on crates/oxc_transformer/src/plugins/tagged_template_transform.rs:
```diff
@@ -0,0 +1,206 @@
+            let raw_bytes = &quasi.value.raw.as_bytes();
+            // Get the bytes up to the last possible starting position of the script tag
+            let max_remain_len = raw_bytes.len().saturating_sub(SCRIPT_TAG.len());
+            let raw_bytes_iter = raw_bytes[..=max_remain_len].iter().copied().enumerate();
```
Aha, fixed now

sapphi-red reviewed (approved):
LGTM on my side

overlookmotel commented:
Actually, this transform is enabled in benchmarks. Effect is only -1%.

That's acceptable for now. We can improve it in a follow-up.

overlookmotel reviewed (approved):
We've decided to merge, and fix the bug + perf in follow-ups.

Dunqing committed:
feat(transformer): support tagged template expression with `</script` transformation (#15664)

Implement a transformation plugin for tagged template expressions containing </script> tags to prevent script tag issues in browser environments.

  This plugin transforms tagged template literals containing `</script` (case-insensitive) to use a helper function call with template caching, matching esbuild's behavior.

 ### Example

  Transforms tagged templates to use cached template objects with a helper function:

  Input:
```js
foo`</script>`
bar`<script>${content}</script>`
```
  Output:
```js
var _templateObject, _templateObject2;
foo(_templateObject || (_templateObject = babelHelpers.taggedTemplateLiteral(["<\/script>"])));
bar(_templateObject2 || (_templateObject2 = babelHelpers.taggedTemplateLiteral(["<script>", "<\/script>"])), content);
```

 Closes #15306
